adapters/src/main/java/io/realm/RealmCursor.java

```diff
@@ -0,0 +1,688 @@
+    public boolean move(int offset) {
+        return moveToPosition(position + offset);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean moveToPosition(int newPosition) {
+        checkClosed();
+        if (newPosition < -1) {
+            position = -1;
+            return false;
+        } else if (newPosition > getCount()) {
+            position = getCount();
+            return false;
+        } else {
+            position = newPosition;
+            return true;
+        }
+    }
+
+    private void checkClosed() {
+        if (closed)
+            throw new IllegalStateException("Cursor is closed");
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean moveToFirst() {
+        return moveToPosition(0);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean moveToLast() {
+        return moveToPosition(getCount() - 1);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean moveToNext() {
+        return move(1);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean moveToPrevious() {
+        return move(-1);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isFirst() {
+        return getCount() > 0 && getPosition() == 0;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isLast() {
+        int count = getCount();
+        return count > 0 && position == count - 1;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isBeforeFirst() {
+        return getCount() == 0 || position == -1;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isAfterLast() {
+        int count = getCount();
+        return count == 0 || position == count;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int getColumnIndex(@NonNull String columnName) {
+        if (data != null) {
+            int columnIndex = (int) data.getTable().getColumnIndex(columnName);
+            return columnIndex == TableOrView.NO_MATCH ? -1 : columnIndex;
+        } else {
+            return -1;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int getColumnIndexOrThrow(@NonNull String columnName) throws IllegalArgumentException {
+        checkClosed();
+        if (data == null) {
+            throw new IllegalStateException("data is null");
+        }
+
+        int columnIndex = (int) data.getTable().getColumnIndex(columnName);
+
+        if (columnIndex == TableOrView.NO_MATCH) {
+            throw new IllegalArgumentException("column not found");
+        }
+
+        return columnIndex;
+
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    @Nullable
+    public String getColumnName(int columnIndex) {
+        checkClosed();
+        if (data != null) {
+            return data.getTable().getColumnName(columnIndex);
```

Comment:
I would prefer if this went through the schema instead, since it would apply any additional rules we might apply for names.